dockercompose v2

ITJamie opened the pull request:
### What does this PR do?
add python-on-whales support for docker

### What issues does this PR fix or reference?
Fixes: #62743

### Previous Behavior
docker-compose v1 (python lib) only supported

### New Behavior
adds docker-compose v2 support (via python-on-whales library)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

jappi00 commented:
Hello @ITJamie,

Do you need any help here? I would love to have an acceptable solution for docker compose.

Thank you very much!

jappi00 commented on salt/modules/dockercompose.py:
```diff
@@ -330,13 +340,16 @@ def __load_project_from_file_path(file_path):
-    except Exception as inst:  # pylint: disable=broad-except
-        return __handle_except(inst)
+    if HAS_PYTHON_ON_WHALES:
+        project = DockerClient(compose_files=[file_path])
```
Maybe we should consider later allowing multiple docker compose files.

ITJamie replied:
out of interest what would this gain. the module calls happen right now based on a specific filename / folder thats being called. 
that would seem like a change from existing functionality. 

my goal for this PR is to replicate existing functions only. that makes this easier to get merged in / tested / approved as a deprecation of the existing import

jappi00 replied:
Yeah, I only wanted to point it out. I needed the functionality in the past. But youre right that is more a topic for an own PR / Issue. (https://docs.docker.com/compose/extends/)

ITJamie commented:
Quite happy for more suggestions or to add you to the branch. Ive been a bit busy with day to day. 

The current changes make it functional in almost the same way as previous docker module. Create function is the only one that differs. 

Right now python-on-whales doesn't have a debug output option so that's technically a bit of a change

jappi00 commented on salt/modules/dockercompose.py:
```diff
@@ -915,8 +960,11 @@ def up(path, service_names=None):
-            result = _get_convergence_plans(project, service_names)
-            ret = project.up(service_names)
+            if HAS_PYTHON_ON_WHALES:
+                project.compose.up(services=service_names)
```
When I try this, it runs indefinitely. I am uncertain what it was before. I think it needs to be run in detached mode.
`project.compose.up(services=service_names, detach=True)`

ITJamie commented on salt/modules/dockercompose.py:
```diff
@@ -475,6 +488,7 @@ def create(path, docker_compose):
 
         salt myminion dockercompose.create /path/where/docker-compose/stored content
     """
+    # TODO: needs adjustment for python on whales
```
This function (create) is not doing what `docker compose create` should be doing. This seems like an antipattern as every other compose function listed is a 1-1 cli command match.

as such i have not touched it yet as i feel we need community input

Ch3LL replied:
Can you document this so users are aware when using the python on whales library.

ITJamie replied:
ive added a new function create_command. which actually runs the create command in docker-compose.

in the longterm the existing "create" command should probably become "create_file" or something similar.

ITJamie commented:
Minor blockage from: https://github.com/gabrieldemarmiesse/python-on-whales/issues/395

The current python-on-whales compose functions do not return output of successful commands (it can send it to stdout, which isn't great).

Because of this i cant 100% match existing functionality when users use the "debug" flag when running the docker compose module.

Otherwise the "main" docker-compose functions are done. (other than create, due to the fact the create function in the salt docker module doesnt do what the create command in docker compose cli does)

ITJamie commented:
@whytewolf I somehow removed you as a reviewer on this. Can you add yourself or someone on the salt team back on this?

Ch3LL commented on salt/modules/dockercompose.py:
```diff
@@ -583,12 +606,16 @@ def restart(path, service_names=None):
     project = __load_project(path)
     debug_ret = {}
     result = {}
+    # TODO: needs work
```
Can this be removed?

Ch3LL commented on salt/modules/dockercompose.py:
```diff
@@ -583,12 +606,16 @@ def restart(path, service_names=None):
+                project.compose.restart(services=service_names, quiet=True)
+            else:
+                project.restart(service_names)
+            if debug and not HAS_PYTHON_ON_WHALES:
```
Can we add debug functionality for the python on whales library?

What are we adding todo's for this functionality? 

If we decide that it needs to be implemented later we need to atleast document it for each function that debug does not work when using python on whales.

ITJamie replied:
I have just completed this for every command

Ch3LL reviewed (changes requested):
This will require:
- a changelog
- test coverage (you can look at tests/pytests/functional/states/test_docker_container.py for inspiration)
- add the new library to requirements/static/ci/linux.in so the tests are testing the new library.

ITJamie commented:
ive added all of the debug outputs as per the original code. it should now be ready for review

ITJamie commented:
re-run pr-photon-3-x86_64-py3-pytest
re-run pr-alma-8-x86_64-py3-pytest

dwoz commented:
@ITJamie Is this something we can get into 3008.x ?

ITJamie commented:
Honestly at this point i would have imagined i would recreate this pr against salt-ext-docker when the repo was made.

twangboy commented on salt/modules/dockercompose.py:
```diff
@@ -0,0 +1,1202 @@
+
+    HAS_PYTHON_ON_WHALES = True
+except ImportError:
+    HAS_PYTHON_ON_WHALES = False
```
Problem:
Importing DockerClient from python_on_whales at the global module level introduces dangerous side effects. When python-on-whales is initialized during a module import, it immediately spawns background subprocesses to discover the local system's Docker and Docker Compose binary versions (e.g., executing docker version). Because Salt's loader scans and imports every execution module during minion startup and refresh cycles, this tracking behavior will trigger globally across the entire fleet—including on minions without Docker installed, non-Linux platforms, or environments with restricted execution privileges, causing performance degradation or unhandled loader crashes.

Suggested Change:
Use importlib.util.find_spec to verify the library's presence in a completely side-effect-free manner at the global level, allowing you to defer the intensive client import into local functions.

```python
import importlib.util

try:
    HAS_PYTHON_ON_WHALES = importlib.util.find_spec("python_on_whales") is not None
except ImportError:
    HAS_PYTHON_ON_WHALES = False
```

twangboy commented on salt/modules/dockercompose.py:
```diff
@@ -0,0 +1,1202 @@
+    """
+    if HAS_PYTHON_ON_WHALES:
+        project = DockerClient(compose_files=[file_path])
+    else:
```
Problem:
Directly instantiating DockerClient here relies on the global import that must be removed to protect the Salt loader. To make the module safe and stable across the ecosystem, heavy third-party dependencies and their downstream orchestrations should be imported lazily right where they are consumed.

Suggested Change:
Pull the DockerClient import inline inside the execution path block so it only runs on minions actively executing a Docker Compose task.

```python
    if HAS_PYTHON_ON_WHALES:
        from python_on_whales import DockerClient
        project = DockerClient(compose_files=[file_path])
    else:
```

ITJamie commented:
requested changes added

ITJamie committed:
Rewrite dockercompose module to use python-on-whales

Replace docker-compose v1 (EOL) with python-on-whales as the backend.
Squashed from PR #63051 (author: ITJamie).

Refs #62743

dwoz committed:
Gate python_on_whales backend behind opt-in config for LTS

Add a `dockercompose:use_python_on_whales` minion config flag that selects
the `python_on_whales` (docker compose v2) backend. On 3006.x/3007.x/3008.x
the legacy `compose` library remains the default so existing users are not
broken by the new dependency or by any behavioural differences between the
two implementations. If the flag is set but `python_on_whales` is not
installed, log a warning and fall back to the legacy backend rather than
failing.

The default flips to `python_on_whales` in 3009, at which point the flag
becomes a deprecated no-op.

Also swap `project.ps(all=True)` for `project.compose.ps(all=True)` in the
v2 branch of the lifecycle helpers so the container listing is compose-
scoped rather than host-scoped.

Consolidate the two changelog entries (62743, 63051) into a single 63051
entry describing the opt-in behaviour.
